server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -312,7 +314,21 @@ public Query regexpQuery(String value, int flags, int maxDeterminizedStates, @Nu
             + "] which is of type [" + typeName() + "]");
     }
 
-    public abstract Query existsQuery(QueryShardContext context);
+    /**
+     * Create a query that matches any document that contains this field.
+     * The default implementation works for most of the field types and creates a {@link DocValuesFieldExistsQuery} when doc_values
+     * are available, otherwise a {@link NormsFieldExistsQuery} when norms are available, and in the worst case scenario it consists
+     * of a {@link TermQuery} against the {@link FieldNamesFieldMapper}.
+     */
+    public Query existsQuery(QueryShardContext context) {
+        if (hasDocValues()) {
+            return new DocValuesFieldExistsQuery(name());
+        } else if (omitNorms()) {
```

Comment:
I believe it should be `omitNorms() || indexOptions() != IndexOptions.NONE`

Reply:
I gave up on this, leaving the norms based impl where they are (it's only 3 of them). The standard impl only switches between doc_values and field_names